packages/next/next-server/lib/router/router.ts

```diff
@@ -393,31 +393,60 @@ export default class Router implements BaseRouter {
 
       // If shallow is true and the route exists in the router cache we reuse the previous result
       this.getRouteInfo(route, pathname, query, as, shallow).then(routeInfo => {
-        const { error } = routeInfo
+        const doRouteChange = (routeInfo: RouteInfo, emit: boolean) => {
+          const { error } = routeInfo
 
-        if (error && error.cancelled) {
-          return resolve(false)
-        }
+          if (error && error.cancelled) {
+            return resolve(false)
+          }
 
-        Router.events.emit('beforeHistoryChange', as)
-        this.changeState(method, url, addBasePath(as), options)
+          if (emit) {
+            Router.events.emit('beforeHistoryChange', as)
+            this.changeState(method, url, addBasePath(as), options)
 
-        if (process.env.NODE_ENV !== 'production') {
-          const appComp: any = this.components['/_app'].Component
-          ;(window as any).next.isPrerendered =
-            appComp.getInitialProps === appComp.origGetInitialProps &&
-            !(routeInfo.Component as any).getInitialProps
-        }
+            if (process.env.NODE_ENV !== 'production') {
+              const appComp: any = this.components['/_app'].Component
+              ;(window as any).next.isPrerendered =
+                appComp.getInitialProps === appComp.origGetInitialProps &&
+                !(routeInfo.Component as any).getInitialProps
+            }
+          }
 
-        this.set(route, pathname, query, as, routeInfo)
+          this.set(route, pathname, query, as, routeInfo)
 
-        if (error) {
-          Router.events.emit('routeChangeError', error, as)
-          throw error
+          if (emit) {
+            if (error) {
+              Router.events.emit('routeChangeError', error, as)
+              throw error
+            }
+
+            Router.events.emit('routeChangeComplete', as)
+            resolve(true)
+          }
         }
 
-        Router.events.emit('routeChangeComplete', as)
-        return resolve(true)
+        if ((routeInfo as any).dataRes) {
+          const dataRes = (routeInfo as any).dataRes as Promise<RouteInfo>
+          // race loading state timeout with data response
+
+          Promise.race([
+            new Promise(resolve => setTimeout(() => resolve(false), 110)),
```

Comment:
This should have a comment

Reply:
Good point, updated the comment to be more verbose explaining why we're racing the two